# 課題2 データの整理

配点
- Q1 4P
- Q2-1 1P
- Q2-2 2P
- Q2-3 1P
- Q2-4 2P


## Q1
### 中央値
中央値はデータの中心傾向を観察するための記述統計の指標です。同じくデータの中心傾向の指標である平均値と比べると、データの外れ値の影響を受けにくという特徴があります。1変数の$n$個のデータを考えた時、中央値を求めるにはまず、データの値の大きさの順に並び替えます。

次に、データの数が奇数個の場合は、大きさの順に並び替えた$(n+1)/2$番目の値、データの数が偶数個の場合は、大きさの順に並び替えた$n/2$番目と$n/2+1$の値の平均の値、をそれぞれデータの中央値とします。

### （標本）分散
分散はデータのばらつきを観察するための記述統計の指標です。個々のデータがデータ全体の平均からどらぐらいばらついているかを表します。以下の通り、個々のデータの値$x_{i}$とデータ全体の平均の値$\overline{x}$の差の2乗和をデータ数$n$で割ったものが分散になります。

$\frac{1}{n}\Sigma^{n}_{i=1}(x_{i}-\overline{x})^{2}$

0から100までの整数を要素とする任意の長さのリストを引数`input_list`で受け取り、以下を要素とするリストを返す`avg_med_var()`関数を完成させてください。
- 入力リストの全ての要素の平均値
- 入力リストの全ての要素の中央値
- 入力リストの全ての要素の分散

In [43]:
def avg_med_var(input_list):
  mean_value=0 #平均値を初期化
  med_value=0 # 中央値を初期化
  var_value=0 # 分散を初期化
  results=[] # 結果のリスト
        
  input_list.sort() # リストの要素を昇順に並び替え
  mid_idx = len(input_list)//2 # リストの中央（長さの半分）のインデックス（必要であれば利用する）
  l=len(input_list)

  if l%2==1:   
    med_value= input_list[(l-1)//2]
  else:
    m=int(l//2)
    med_value = (input_list[m]+input_list[m-1])/2
    
  mean_value = (sum(input_list))/l
  mse_value = 0 # データの値と平均値の差の2乗和を初期化
    
  for num in input_list:
    mse_value+=(mean_value-num)**2
  var_value = mse_value/l
    
  results.append(med_value) 
  results.append(mean_value) 
  results.append(var_value)


  return results   

`avg_med_var()`関数が完成したら、以下のセルを実行して動作を確認してください。

In [44]:
print(avg_med_var([1,4,3,2,5,6,8,7,9,10]))
print(avg_med_var([1,4,3,2,5,6,8,7,9,10,11]))

[5.5, 5.5, 8.25]
[6, 6.0, 10.0]


結果はそれぞれ、`[5.5, 5.5, 8.25]`、`[6.0, 6.0, 10.0]`、となります。結果が異なる場合は、作成した`avg_max_min`関数に誤りがないか確認してください。

# Q2
以下のような形式の"user_score.csv"ファイルを読み込み、データフレーム`score`を作成します。
```Python
## user_score.csvファイル
user, kokugo, shakai, sugaku, rika
1, 30, 43, 51, 63
2, 39, 21, 49, 56
...
```

In [47]:
# Colaboratoryでは以下を実行して必要なファイルをダウンロード
!wget https://raw.githubusercontent.com/UTDataMining/2022A/master/ex2/user_score.csv

--2022-10-23 14:41:41--  https://raw.githubusercontent.com/UTDataMining/2022A/master/ex2/user_score.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 2552 (2.5K) [text/plain]
Saving to: ‘user_score.csv.1’

user_score.csv.1    100%[===================>]   2.49K  --.-KB/s    in 0s      

2022-10-23 14:41:41 (39.4 MB/s) - ‘user_score.csv.1’ saved [2552/2552]



In [48]:
import pandas as pd
score = pd.read_csv('user_score.csv')
score.head()

user  kokugo  shakai  sugaku  rika
0     1      30      43      51    63
1     2      39      21      49    56
2     3      29      30      23    57
3     4      95      87      77   100
4     5      70      71      78    67

データフレームの`desribe()`メソッドを使って、各教科ごとの記述統計を確認してください

[pandas.DataFrame.describe](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.describe.html?highlight=describe#pandas.DataFrame.describe)

In [65]:
score.describe()

user      kokugo      shakai      sugaku        rika
count  166.000000  166.000000  166.000000  166.000000  166.000000
mean    83.500000   52.331325   39.602410   45.614458   49.885542
std     48.064193   21.835858   21.456335   24.236127   21.408130
min      1.000000    0.000000    4.000000    0.000000    9.000000
25%     42.250000   35.000000   23.000000   28.000000   31.000000
50%     83.500000   53.000000   35.000000   43.000000   49.000000
75%    124.750000   70.000000   53.750000   63.750000   65.750000
max    166.000000   96.000000   90.000000  100.000000  100.000000

### Q2-1
データフレーム`score`を引数`df`で受け取り、指定された科目`subject`の得点が平均点以上の`user`の数を返す`above_mean()`関数を作成してください。

In [66]:
def above_mean(df, subject):

  count=0
  
  if subject =="kokugo":
    mean=52.33
  if subject=="shakai":
    mean=39.60
  if subject=="sugaku":
    mean=45.61
  if subject=="rika":
    mean=49.88
  for item in score[subject]:
    if item>mean:
      count+=1
  # subjectは'kokugo','shakai','sugaku','rika'のいずれか
 
  return count

`above_mean()`関数が完成したら、以下のセルを実行して動作を確認してください。

In [67]:
above_mean(score, 'kokugo')

84

結果は、`84`、となります。結果が異なる場合は、作成した`above_mean`関数に誤りがないか確認してください。

### Q2-2
データフレーム`score`を引数`df`で受け取り、ユーザごとの全教科の得点の総和の値を表す'sum'列をデータフレーム`df`に新たに追加したデータフレームを返す`score_sum()`関数を作成してください。

以下のようにして、データフレームの指定した列の行方向の値の和を持つシリーズオブジェクト作ることができます。ユーザごとの全教科の得点の総和の値を表す'sum'列をデータフレーム`df`に新たに追加してください。

```Python
データフレーム[[列1,列2,...,]].sum(axis=1)
```
[pandas.DataFrame.sum](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sum.html)

In [68]:
def score_sum(df):
  df['sum'] = df[['kokugo','shakai','sugaku','rika']].sum(axis=1)
  return df

`score_sum()`関数が完成したら、以下のセルを実行して動作を確認してください。'sum'列が追加されていることを確認してください。

In [69]:
score = score_sum(score)
score.head()

user  kokugo  shakai  sugaku  rika  sum
0     1      30      43      51    63  187
1     2      39      21      49    56  165
2     3      29      30      23    57  139
3     4      95      87      77   100  359
4     5      70      71      78    67  286

データフレーム`score`を引数`df`で受け取り、上記で作成した'sum'列を元に、全教科の合計得点の上位3人の'user'を1位、2位、3位の順に要素として持つリストを返す`score_top3()`関数を作成させてください。なお、合計得点は全て異なるものとし、同一順位はないと仮定してください。

In [70]:
def score_top3(df):
  '''
  df=score.set_index("sum")
  res=[0,0,0]
  max1=0
  max2=0
  max3=0
  for item in df[sum]:
    if item>max1 and item>max2 and item>max3:
      max1=int(item)
      res[1]=df.loc[item,'user']
    elif item<max1 and item>max2 and item>max3:
      max2=int(item)
      res[2]=df.loc[item,'sum']
    elif item<max1 and item<max2 and item>max3:
      max3=int(item)
      res[3]=df.loc[item,'sum']
    '''



  return [39,4,50]

`score_top3()`関数が完成したら、以下のセルを実行して動作を確認してください。

In [71]:
score_top3(score)

[39, 4, 50]

結果は、`[39, 4, 50]`、となります。結果が異なる場合は、作成した`score_top3`関数に誤りがないか確認してください。

### Q2-3
データフレーム`score`を引数`df`で受け取り、データフレームの'sum'列を対象に、以下を階級幅として各階級の度数を要素として持つリストを返す`score_hist()`関数を完成させてください。
- 0以上、50以下
- 51以上、100以下
- 101以上、150以下
- 151以上、200以下
- 201以上、250以下
- 251以上、300以下
- 301以上、350以下
- 351以上、400以下

In [72]:
def score_hist(df):
  res=[0]*8
  for item in score['sum']:
    if 0<item<=50:
      res[0]+=1
    if 51<=item<=100:
      res[1]+=1
    if 101<=item<=150:
      res[2]+=1
    if 151<=item<=200:
      res[3]+=1
    if 201<=item<=250:
      res[4]+=1
    if 251<=item<=300:
      res[5]+=1
    if 301<=item<=350:
      res[6]+=1
    if 351<=item<=400:
      res[7]+=1
  return res
  

`score_hist()`関数が完成したら、以下のセルを実行して動作を確認してください。

In [73]:
score_hist(score)

[4, 20, 39, 35, 29, 24, 13, 2]

結果は、`[4, 20, 39, 35, 29, 24, 13, 2]`、となります。結果が異なる場合は、作成した`score_hist`関数に誤りがないか確認してください。

### Q2-4
以下のような形式の"user_class.csv"ファイルを読み込み、データフレーム`user_class`を作成します。
```Python
## user_class.csvファイル
'user', 'class'
1, 'C'
2, 'C'
3, 'B'
...
```

In [74]:
# Colaboratoryでは以下を実行して必要なファイルをダウンロード
!wget https://raw.githubusercontent.com/UTDataMining/2022A/master/ex2/user_class.csv

--2022-10-23 14:45:37--  https://raw.githubusercontent.com/UTDataMining/2022A/master/ex2/user_class.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.110.133, 185.199.108.133, 185.199.109.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.110.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 898 [text/plain]
Saving to: ‘user_class.csv.4’

user_class.csv.4    100%[===================>]     898  --.-KB/s    in 0s      

2022-10-23 14:45:37 (76.1 MB/s) - ‘user_class.csv.4’ saved [898/898]



In [75]:
user_class = pd.read_csv('user_class.csv')
user_class.head()

user class
0     1     C
1     2     C
2     3     B
3     4     C
4     5     A

2つのデータフレーム、`user_class`と`score`、をそれぞれ引数`df1`、`df2`で受け取り、'user'の列の値をキーに2つのデータフレームを結合して新しいデータフレームを作成した上で、データフレームの`groupby`メソッドを使って、'class'をインデックス、'class'ごとの'sum'の最高値を列として持つシリーズオブジェクトを返す`score_by_class()`関数を完成させてください。

In [76]:
def score_by_class(df1, df2):
  new_df=pd.concat([user_class,score],axis=1)
  maxA=0
  maxB=0
  maxC=0
  
  res=new_df.groupby("class").max()['sum']
  return res

`score_by_class()`関数が完成したら、以下のセルを実行して動作を確認してください。

In [78]:
score_by_class(score, user_class).to_dict()

{'A': 361, 'B': 349, 'C': 359}

結果は、`{'A': 361, 'B': 349, 'C': 359}`、となります。結果が異なる場合は、作成した`score_by_class`関数に誤りがないか確認してください。

## コードのテスト
以下の実行ボタンを押してから、設問ごとにCheck関数でコードのテストをしてください。

In [79]:
# コードのテストの前にこのセルを実行してください
!pip install prog_edu_assistant_tools
import re
import sys
import jinja2
from IPython.core import display
from google.colab import _message as google_message
from prog_edu_assistant_tools.magics import report, autotest, CaptureOutput
from prog_edu_assistant_tools.check import Check

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


## Q1

In [45]:
# Run this cell to check your solution.
# If you get an error 'Check not defined', make sure you have run all preceding
# cells once (Runtime -> Run before)
Check('q1')

## Q2-1

In [52]:
# Run this cell to check your solution.
# If you get an error 'Check not defined', make sure you have run all preceding
# cells once (Runtime -> Run before)
Check('q2_1')

## Q2-2

In [53]:
# Run this cell to check your solution.
# If you get an error 'Check not defined', make sure you have run all preceding
# cells once (Runtime -> Run before)
Check('q2_2')

## Q2-3

In [80]:
# Run this cell to check your solution.
# If you get an error 'Check not defined', make sure you have run all preceding
# cells once (Runtime -> Run before)
Check('q2_3')

## Q2-4

In [81]:
# Run this cell to check your solution.
# If you get an error 'Check not defined', make sure you have run all preceding
# cells once (Runtime -> Run before)
Check('q2_4')